### ISSUE TITLE
Inquiry Regarding KalmanNet for Sensor Fusion in ROS

### ISSUE THREAD

**Issue by jeffchen1108** [SongJgit/KalmanNet4SensorFusion#3]
Dear @SongJgit ,

I hope this message finds you well. My name is Jeff, and I am a student currently conducting research on sensor fusion using IMU and odometry data within the ROS framework. I recently came across your excellent work on KalmanNet, and I am very interested in applying your model in a simulated environment (Gazebo) to achieve more accurate state estimation using only IMU and odometry data. I plan to use your KalmanNet sensor fusion model as a replacement for the EKF (Extended Kalman Filter) that I have been using, as I believe KalmanNet can provide superior sensor fusion capabilities.

I have successfully trained the model using the code provided in your GitHub repository, but I encountered some difficulties when integrating KalmanNet into my ROS system. Specifically, I am unsure about the exact format of the input data when feeding IMU and odometry data into the model. For example, should the IMU input include the x, y, and z-axis accelerations along with the z-axis angular velocity? I noticed that in the train.pt, val.pt, and test.pt files, the IMU data is 4-dimensional, but I am unclear on which specific data should be input to the model.

Additionally, I would like to ask if KalmanNet has been tested in ROS or on actual autonomous vehicles. If so, could you share any insights or suggestions that might help me with the challenges I am currently facing regarding the input data?

Your guidance would be invaluable for my research, and if I successfully apply your work, I would be happy to cite your paper in my future publications. Thank you for your time, and I look forward to your response.

Best regards,
Jeff

**Comment by SongJgit** [comment]
The code comments in that section are written rather messily.
The exact inputs are dependent on your model (F, H).

As you can see from `get_pred_jac` in [make_params.py](https://github.com/SongJgit/KalmanNet4SensorFusion/blob/main/Net/params/make_params.py),  this func gives two parameters, where the first parameter `prev_states` is the `state` maintained by the KF/KalmanNet, and the second parameter `current_states` is the data from the sensor.

For example in my implementation, `prev_states` is always `[[bs, [x, y, x', y', theta_k, omega_k],1]`
IMU: `current_states` is then `[ax , ay , theta, omega]` from IMU. (ax == x''）
Wheel: `current_states`  is `[velocity_left, velocity_right, heading,omega]`

Of course you also change `current_states` to something else like `[x, y, z, x', y', z', x'', y'', z'', theta_k, omega_k, ... ]`, as long as they match your `F/H`.

In fact, if you want to quickly apply this method to your data.
You only need to do the following steps:
1. understand the format of the data and generate your data in the same format, e.g. `training.pt`, see [preprocess.py](https://github.com/SongJgit/KalmanNet4SensorFusion/blob/main/data/NCLT/preprocess.py)
2. modify `get_pred_jac` and `get_obs_jac` in [make_params.py](https://github.com/SongJgit/KalmanNet4SensorFusion/blob/main/Net/params/make_params.py) according to your model
3. Trainning!
4. 
I haven't tested it on ROS or an actual self-driving vehicle.
So can't advise on that.

**Comment by jeffchen1108** [comment]
亲爱的 SongJgit，

非常感谢您之前的详细回复。我现在有一些关于使用 KalmanNet 进行 IMU 和里程计数据融合的额外问题，因为我目前的设置中并不希望使用 GPS。

在阅读 make_params.py 后，我认为我需要创建一个自定义类来负责融合 IMU 和里程计数据。请问我的理解是否正确？我可以通过使用 preprocess.py 生成新的包含 IMU 和里程计数据的 train.pt、val.pt 和 test.pt 文件来使用 NCLT 数据集进行训练吗？然后，我假设只需要修改 train.py 并重新训练模型即可生成一个融合 IMU 和里程计数据的 KalmanNet 模型。我的理解和做法是否正确？

以下是我在 ROS 中可以取用的 IMU 数据：

姿态 (x, y, z, w)
角速度 (x, y, z)
线性加速度 (x, y, z)

里程计数据如下：

位置 (x, y, z)
姿态 (x, y, z, w)
线速度 (x, y, z)
角速度 (x, y, z)
我目前使用 ROS 内置的 EKF 来融合 IMU 和里程计数据。我的最终目标是使用 KalmanNet 的传感器融合来减少传感器的累积误差。我最需要的是模型输出的位置和姿态数据，但如果能同时生成速度和角速度，那将更加理想。
目前我详细的IMU、里程计和经过EKF融合后的数据如以下附上的图片
里程計：
![Screenshot from 2024-09-08 21-28-23](https://github.com/user-attachments/assets/f8fa2e5f-a11e-4b29-8417-4a6f7932acea)
IMU:
![Screenshot from 2024-09-08 21-29-06](https://github.com/user-attachments/assets/2f640e4f-68a0-426d-bfd8-b08f4b1bd63d)
经过EKF融合后的数据：
![Screenshot from 2024-09-08 21-30-06](https://github.com/user-attachments/assets/ee25abca-6f24-4038-a339-a3d14d4801f5)

如果您能确认我的做法是否正确，并提供一些进一步的指导或建议，帮助我实现使用 KalmanNet 进行 IMU 和里程计数据融合，我将非常感激。

再次感谢您的时间与帮助。

此致，
Jeff

**Comment by SongJgit** [comment]
@jeffchen1108 
preprocess.py算是一个示例，你照着它再结合你的数据写预处理，保存的数据结构一致即可。当然可能还需要修改一下dataset，用于读取你的数据。
makeparams里面的类是为了管理不同模型的F和H。
照着它把你的F和H以及他们的jac计算写好就可以了。

不论你的数据是什么样子的，只要ekf可以跑，那么kalmannet就可以跑。

当然，最简单的方案可能是用nclt自己的imu和wheels结合你的model先试试可行性，然后在上你自己的数据。

**Comment by jeffchen1108** [comment]
作者您好我想请问在您的论文中我有看到您在预测阶段使用更新频率高但精度低的感测器进行建模，例如IMU/Wheel Encoders，而校正阶段则对更新频率较低但精度较高的感测器进行建模，例如GPS。并且您提到了即使在缺失GPS的yt情况下也能实现感测器融合，那是不是也代表就算我不去更改您的模型也能实现我对于无GPS情况下的使用呢？也就是说我用nclt数据集的gps、imu、wheel等数据进行训练，而实际使用则不使用gps，这样子模型是有办法做使用的吗？

会这样询问是因为我听取您的建议最近已经重新从nclt上下载imu、wheel等，但不包含gps的数据并制作成我自己的数据集，然而我在更改您的模型与训练档案却遇到很多张量不匹配的问题，我传入的数据与模型的期望输入不匹配以及运算因为张量不匹配而无法进行矩阵运算，我也改了很多档案像是wheel_gpsfusion_origin.py、make_params.py、fusion_trainer. py、nclt_dataset.py、base_trainer.py、fusion_kalman_net.py、kalman_net.py等等，我目前是使用imu的ax,ay,theta跟wheel的vl,vr,omega组成模型的输入，但作者您给的建议说只需要修改make_params.py以及nclt_dataset.py，所以我怀疑是否自己已经改到模型的部份会不会是错误的方向？

**Comment by SongJgit** [comment]
> 作者您好我想请问在您的论文中我有看到您在预测阶段使用更新频率高但精度低的感测器进行建模，例如IMU/Wheel Encoders，而校正阶段则对更新频率较低但精度较高的感测器进行建模，例如GPS。并且您提到了即使在缺失GPS的yt情况下也能实现感测器融合，那是不是也代表就算我不去更改您的模型也能实现我对于无GPS情况下的使用呢？也就是说我用nclt数据集的gps、imu、wheel等数据进行训练，而实际使用则不使用gps，这样子模型是有办法做使用的吗？
> 
> 会这样询问是因为我听取您的建议最近已经重新从nclt上下载imu、wheel等，但不包含gps的数据并制作成我自己的数据集，然而我在更改您的模型与训练档案却遇到很多张量不匹配的问题，我传入的数据与模型的期望输入不匹配以及运算因为张量不匹配而无法进行矩阵运算，我也改了很多档案像是wheel_gpsfusion_origin.py、make_params.py、fusion_trainer. py、nclt_dataset.py、base_trainer.py、fusion_kalman_net.py、kalman_net.py等等，我目前是使用imu的ax,ay,theta跟wheel的vl,vr,omega组成模型的输入，但作者您给的建议说只需要修改make_params.py以及nclt_dataset.py，所以我怀疑是否自己已经改到模型的部份会不会是错误的方向？

1. 我没有尝试过训练好之后不用GPS的情况。但是本身不论是wheel还是IMU，他们都会随着时间而飘移，导致误差越来越大， 这是传感器本身决定的，不会因为用了神经网络就改变(神经网络可能能辅助降低误差，但是不可能做到避免)。

2. 从我的代码架构来说，对于代码部分，的确时只需要修改那两个部分。但是对于网络而言，你需要修改部分参数，比如state_dim和 mask等，这点可以通过更改配置文件做到。
比如：

    2.1 在https://github.com/SongJgit/KalmanNet4SensorFusion/blob/b4b28dd3e0aa5b28c7e2d90fd910eff7512e70c5/configs/nclt/fusion/wheel_gpsfusion_split.py#L19
    
    至少你需要正确的state_dim，这里的state_dim=6对应的就是`[[bs, [x, y, x', y', theta_k, omega_k],1]`，`[x, y, x', y', theta_k, omega_k]`正好6维，这里的observation_axis没有特别的意义。
    ```python
    params = dict(type=NCLTFusionWheelGPSParams,
                  T=1,
                  constant_noise=True,
                  state_dim=6,
                  observation_axis=[0, 3],
                  noise_q2=1,
                  noise_r2=1)
    ```
    
    2.2  在https://github.com/SongJgit/KalmanNet4SensorFusion/blob/b4b28dd3e0aa5b28c7e2d90fd910eff7512e70c5/configs/nclt/fusion/wheel_gpsfusion_split.py#L26
    pred_metric_mask是指的从模型的结果中选取哪些数据去计算metric，比如对`[[bs, [x, y, x', y', theta_k, omega_k],1]`类型的数据的第二维使用`[True, True, False, False, False, False]`，就是选择其中的x，y。
    target_metric_mask是指的从标签数据中选取哪些数据计算metric，由于本身给出的标签就是`[[bs, [x, y],1]`，所以[True, True]选出来的就是x，y
    ```python
    metric = dict(
        pred_metric_mask=[True, True, False, False, False, False],
        target_metric_mask=[True, True],  # must match state
        # number of state to compute metric ,metric_dims must be equal to number of mask'True
        num_metric_dims=2,
        metric_dim_names=['X', 'Y'],
        inverse_transforms=False)
    ```

3. 注意在模型里面所有的数据的形状都是`[bs, [num_state], 1]` 或者`[bs, [num_obs], 1]`

**Comment by SongJgit** [comment]
关于KalmanNet，我这有一个微信讨论组(由不同学校的硕博组成，大家方向有相似性，但不完全一致)，你有加入的想法吗？

**Comment by jeffchen1108** [comment]
我有兴趣，感谢您的回覆
您提到的第一点我能理解，只是原本的传感器融合是使用EKF，但看到您使用KalmanNet的效果后就希望能够使用来取代EKF，来解决EKF使用上的一些缺点，累积误差的部分会使用其他方法想办法解决

**Comment by SongJgit** [comment]
> 我有兴趣，感谢您的回覆 您提到的第一点我能理解，只是原本的传感器融合是使用EKF，但看到您使用KalmanNet的效果后就希望能够使用来取代EKF，来解决EKF使用上的一些缺点，累积误差的部分会使用其他方法想办法解决

给我你的邮箱吧，我把我的微信id发你邮箱

**Comment by jeffchen1108** [comment]
jeffchen891108@gmail.com

**Comment by SongJgit** [comment]
> [jeffchen891108@gmail.com](mailto:jeffchen891108@gmail.com)

wx id已经发你邮箱了

**Comment by jeffchen1108** [comment]
我目前使用filtered_wheel作为融合IMU和Wheel的correction，我使用作者您提供的数据集test.pt、train.pt和val.pt有成功。
但是我想取得更多数据，所以我又从nclt网站下载所有日期的感测器数据并重新制作一个数据集，但是因为preprocess.py里面的filtered_wheel处理逻辑
if wheel_counter != prev_wheel_counter:
 filtered_wheel.append(torch.tensor([v_left_wheel[wheel_counter], v_right_wheel[wheel_counter]]))
 prev_wheel_counter = wheel_counter
 else:
 filtered_wheel.append(torch.tensor([torch.nan, torch.nan]))
让我制作的数据集里的filtered_wheel产生了很多的NaN值，这让我的训练在第一个validation就出现了Inf detected in m1x_posterior的错误
但我去检查作者您提供的数据集却没有在filtered_wheel发现任何的NaN
想请问您是怎么处理原本会产生的NaN值的？是有再做额外的预处理吗？

**Comment by SongJgit** [comment]
> 我目前使用filtered_wheel作为融合IMU和Wheel的correction，我使用作者您提供的数据集test.pt、train.pt和val.pt有成功。 但是我想取得更多数据，所以我又从nclt网站下载所有日期的感测器数据并重新制作一个数据集，但是因为preprocess.py里面的filtered_wheel处理逻辑 if wheel_counter != prev_wheel_counter: filtered_wheel.append(torch.tensor([v_left_wheel[wheel_counter], v_right_wheel[wheel_counter]])) prev_wheel_counter = wheel_counter else: filtered_wheel.append(torch.tensor([torch.nan, torch.nan])) 让我制作的数据集里的filtered_wheel产生了很多的NaN值，这让我的训练在第一个validation就出现了Inf detected in m1x_posterior的错误 但我去检查作者您提供的数据集却没有在filtered_wheel发现任何的NaN值 想请问您是怎么处理原本從nclt下載並經過preprocess.py会在filtered_wheel产生的NaN值的？是有再做额外的预处理吗？

两者的更新频率是不一致的，在我的论文实验中wheel是要快于gps的所以在没有gps的数据的时刻我都是使用nan填充，这个在preprocess是可以看到的。
然后在模型代码中，可以看到我是有对量测值进行检查的，如果是nan表明当前时刻是没有量测值也就是gps数据的，会有额外的处理逻辑。

**Comment by jeffchen1108** [comment]
了解我会再多进行尝试，我有另一个疑问是
在我查看 KalmanNet 和 Split-KalmanNet 的配置时，我发现：

在 原始 KalmanNet 中，你们使用了 train_seq_len=50，slide_win_size=4，detach_step=2，这似乎引入了滑动窗口和频繁的梯度分离机制。
而在 Split-KalmanNet 中，你们使用了 train_seq_len=100，slide_win_size=100，detach_step=100，这种配置看起来像是没有采用滑动窗口和频繁的梯度分离。
根据我的理解，论文中的 Split-KalmanNet 应该是对 KalmanNet 进行了一些改进，特别是在长时间序列上使用了梯度分离和滑动窗口机制。然而，配置文件中的设置似乎与我的理解有所不同。原始 KalmanNet 的配置看起来更像是应用了 Split-KalmanNet 的思想，而 Split-KalmanNet 的配置则更像是传统的 KalmanNet。

我想请教以下问题：

为什么原始 KalmanNet 的配置中也使用了滑动窗口和梯度分离机制？这样是否有助于提高模型的稳定性？
为什么 Split-KalmanNet 的配置中采用了与 KalmanNet 类似的设置，而没有使用滑动窗口和频繁的梯度分离？
另外，我还注意到：

原始 KalmanNet 使用的是 MSE 作为损失函数；
Split-KalmanNet 使用的是 SmoothL1 作为损失函数。
因此，我还想请教一下： 3. 选择不同损失函数的原因是什么？为什么原始 KalmanNet 使用 MSE，而 Split-KalmanNet 使用 SmoothL1？这是否与模型对误差或异常值的处理方式相关？

期待能够听到您的解释，再次感谢您们的出色工作！
